### ISSUE TITLE
added uv_poll_init_scoket for windows

### ISSUE THREAD

**Issue by museghost** [nats-io/cnats#98]
Dear NATS team

In the adapters/libuv.h, below changes would be needed since the adapter cannot attach the event loop in the window. Also my windows environment consists of Mingw64 and libuv 1.14.0.

error: 
```bash
Error: 1 - Error - (conn.c:1484): Error attaching to the event loop: 1 - Error
Stack: (library version: 1.7.0)
  01 - _processConnInit
  02 - _connect
  03 - natsConnection_Connect
```

workaround:
```cpp
#if defined(_WIN32) || defined(_WIN64) || defined(__MINGW32__) || defined(__MINGW64__)
    if ((s == NATS_OK)
        && (uv_poll_init_socket(nle->loop, nle->handle, nle->socket) != 0))
#else
    if ((s == NATS_OK)
        && (uv_poll_init(nle->loop, nle->handle, nle->socket) != 0))
#endif
```

I assume this issue is caused by the type of Windows socket might be not as same as the one of the linux or unix.

After investigating this, if yes, could you please apply the above changes as well ?

Thank you for your continuous development

**Comment by kozlovic** [comment]
@museghost Are you using source that include this merged change to attach? https://github.com/nats-io/cnats/pull/93
The library calls `natsLibuv_Attach` with `natsSock` which on Windows is a SOCKET and not an int. So this should work.

Note that in libuv master, the call `uv_pool_init_socket` has been removed (see https://github.com/libuv/libuv/pull/1166). So your changes would work only in the 1.x branch.

Let's first make sure that you are using the changes described above.
Thanks!

**Comment by museghost** [comment]
@kozlovic 
At first, thank you for your guide. For both two points, here is my situation. 

1) the source what I have used is the latest one after #93 patch already. 


2) For being ready for 2.x branch, hmm... I need to more investigate and test it. 

As you see the source of libuv 1.14 branch around uv_poll_init in src/win/poll.c (since line 508), uv_poll_init tries to get the os file handler using uv__os_osfhandler().

And the function might make this issue and that's what I want to check deeply. Then I will let you know it asap. 

Also my windows environment is windows server 2012 R2 64 bit with mingw 64 7.1.0. 

Thank you again.

**Comment by museghost** [comment]
@kozlovic 
Dear CNATS team

For this issue, it is found at https://github.com/libuv/libuv/issues/1077. According to the libuv dev team, they guided as below.

> On Windows uv_poll_init should be used for CRT fds and uv_poll_init_socket for SOCKET handles.

If natsSocket  (in fact, ctx->fd) is SOCKET in windows, uv_poll_init_socket() is required as above, however, for 2.0 branch, it might be not feasible.

Could you please tell me your opinion on that ? Do we need a patch ?


Best Regards
Thank you again,

**Comment by kozlovic** [comment]
@museghost Sorry for the delay. Could you try with code from PR #117?

I am using `UV_VERSION_MAJOR` to detect if running against libuv 1.x or above. As you know, starting in 2.0, they have removed `uv_poll_init_socket` and actually renamed it to `uv_poll_init`, so when running with libuv 2.0,  we have to call `uv_poll_init` (since the `xx_socket` API has been removed) and `uv_poll_init_socket` on libuv < 2.0.

I have checked and it works for me on MacOS X (tried with libuv 1.16 and master) and Windows (tried with 2.0). Please let me know if it works for you and I will merge it to master.

**Comment by kozlovic** [comment]
@museghost I did not get confirmation that this worked for you, but I went ahead and merged the PR #117. Hope this helps. Thanks!
